### DISCUSSION TITLE
Raster Mesh to VTK

### DISCUSSION THREAD

**Discussion by hem-28** [kip-hart/MicroStructPy#73] [General]
Hello,

I am familiar with the raster mesh .txt format written out by MicroStructPy. Pardon my ignorance for that. I wanted to convert this Voxelated data to a VTK file which is similar to the output from Dream3D, so I wanted some clarification. 
1. Does 'Element Attributes' correspond to the grain number/seed number? 
2. In which order is this list written? In this order probably? 
      for (i=0; i<nx; i++)
          for (j=0; j<ny; j++)
            for (k=0; k<nz; k++)
3. The phase/material information is printed in 'rastermesh.png', but this information is not available in rastermesh.txt right? Do I have to correlate the grain id to phase id using the seeds.txt file or is there an alternate direct solution? 
Thanks!

**Comment by hem-28** [comment]
Okay so I figured out  that it is possible to export the raster mesh as vtk (including phase IDs) using the python version. The data is however written out as an unstructured grid format as points data, which is a bit surprising for a regular grid rastermesh. Would there be any possibility to export vtk as Structured grid and cell data ?

**Comment by kip-hart** [comment]
Hi @hem-28 , MicrostructPy was initially written to generate unstructured meshes so there may have been some re-use of the export functions when I added the RasterMesh class. The VTK format section of RasterMesh.write is:

https://github.com/kip-hart/MicroStructPy/blob/44b6a65960b030ac889e7ed873dca9ecac3fd2e2/src/microstructpy/meshing/trimesh.py#L1165-L1216

I'll dig into the VTK file format docs and see if I can use the structured grid type.

**Reply by kip-hart** [reply]
Now that I'm reading the docs from VTK (http://www.princeton.edu/~efeibush/viscourse/vtk.pdf) it looks like the `RECTILINEAR_GRID` option would be better suited than `STRUCTURED_GRID`.

**Reply by hem-28** [reply]
Yes True , Rectilinear_grid would be a better option.

**Comment by kip-hart** [comment]
Hi @hem-28 I have just opened PR #75 to make this modification. I'll let the tests run over night and update the docs tomorrow. New release with the update should be available in the next couple of days.

**Reply by hem-28** [reply]
@kip-hart Thanks for taking it as a feature request!

**Comment by kip-hart** [comment]
Hey @hem-28, I included this feature as of version 1.5.5, released a few minutes ago. Try `pip install --upgrade microstructpy` and see if that works for you. Let me know how it goes, I'm here to help.

**Comment by hem-28** [comment]
Hi @kip-hart . Do you probably have your test script? I tried with mine , and I get the following error. I use Python 3.9.7
 `File "/isi/w/jana/User_Softwares/Anaconda_2021/lib/python3.9/site-packages/microstructpy/meshing/trimesh.py", line 1274, in write
  e_ind = np.where(np.all(elems == el, axis=1))[0][0]
  IndexError: index 0 is out of bounds for axis 0 with size 0`

This is my script.


      <input>
         <material>
             <fraction> 1 </fraction>
             <volume>
                 <dist_type> lognorm </dist_type>
                 <scale> 1 </scale>
                 <s> 0.95 </s>
             </volume>
             <shape> sphere </shape>
             <name> Phase 1 </name>
         </material>
     
         <material>
             <fraction> 3 </fraction>
             <volume>
                 <dist_type> lognorm </dist_type>
                 <scale> 0.5 </scale>
                 <s> 1.01 </s>
             </volume>
             <name> Phase 2 </name>
         </material>
     
         <domain>
             <shape> cube </shape>
             <side_length> 7 </side_length>
             <corner> (0, 0, 0) </corner>
         </domain>
     
         <settings>
             <directory> Seeds </directory>
             <verbose> True </verbose>
             <filetypes>
                 <seeds> vtk </seeds>
                 <poly> vtk </poly>
     			<tri> vtk </tri>
                 <seeds_plot> png </seeds_plot>
                 <poly_plot> png </poly_plot>
     			<tri_plot> png </tri_plot>
                 <verify_plot> png </verify_plot>
             </filetypes>
             <mesher> raster </mesher>
     		<mesh_size> 0.5 </mesh_size>  
             <seeds_kwargs>
                 <linewidths> 0.2 </linewidths>
             </seeds_kwargs>
             <poly_kwargs>
                 <linewidths> 0.2 </linewidths>
             </poly_kwargs>
         </settings>
     </input>

**Reply by hem-28** [reply]
I installed MicroStructPy in a new conda environment, including dependencies. But still end with the same error.

**Reply by hem-28** [reply]
To resolve the issue, I did the following changes to the write function in RasterMesh class :


            coords = [list(set(ax)) for ax in zip(*self.points)]
            coords = np.sort(coords)   # This was important, else the data was not in ascending order, ParaView threw up an error
            .
            .
            .
            el = np.where(mask)
            try:
                  e_ind = np.where(np.all(elems == el, axis=1))[0][0]
                  .
                  .
                  .
                 # PhasesInfo
            except :
                  continue
            with open(filename, 'w') as file:
            file.write(vtk)

**Reply by kip-hart** [reply]
ah good catch, looks like I used too many built-in functions. I replaced those line in this commit: 

https://github.com/kip-hart/MicroStructPy/blob/4937a6918c148b04b67bd18d17930d1cbac17f49/src/microstructpy/meshing/trimesh.py#L1173-L1174

if the CI pipelines pass I'll push an updated version tonight. Sorry for the error!

**Reply by kip-hart** [reply]
Hopefully v1.5.6 will solve the problem. Let me know if it doesn't, the script works fine for me.

**Reply by hem-28** [reply]
Hi @kip-hart Yes it works fine in the new version. Thanks again for the new feature!
